ironfish/src/wallet/scanner/walletScanner.ts

```diff
@@ -0,0 +1,247 @@
+import { DecryptedNote } from '../../workerPool/tasks/decryptNotes'
+import { HeadValue } from '../walletdb/headValue'
+import {
+  RemoteChainProcessor,
+  WalletBlockHeader,
+  WalletBlockTransaction,
+} from './remoteChainProcessor'
+import { ScanState } from './scanState'
+
+export class WalletScanner {
+  readonly logger: Logger
+  readonly nodeClient: RpcClient | null
+  readonly wallet: Wallet
+
+  maxQueueSize: number
+  state: ScanState | null = null
+  lock: Mutex = new Mutex()
+
+  constructor(options: {
+    logger: Logger
+    nodeClient: RpcClient | null
+    wallet: Wallet
+    maxQueueSize: number
+    config: Config
+  }) {
+    this.logger = options.logger
+    this.nodeClient = options.nodeClient
+    this.wallet = options.wallet
+    this.maxQueueSize = options.maxQueueSize
+  }
+
+  get running(): boolean {
+    return !!this.state
+  }
+
+  async abort(): Promise<void> {
+    if (this.state) {
+      await this.state.abort()
+    }
+  }
+
+  async wait(): Promise<void> {
+    if (this.state) {
+      await this.state.wait()
+    }
+  }
+
+  async scan({
+    start,
+    end,
+    force,
+  }: {
+    start?: HeadValue | null
+    end?: HeadValue | null
+    force?: boolean
+  } = {}): Promise<ScanState | null> {
+    if (this.wallet.listAccounts().length === 0) {
+      return null
+    }
+
+    if (this.running && !force) {
+      this.logger.debug('Skipping Scan, already scanning.')
+      return null
+    }
+
+    if (this.running && force) {
+      this.logger.debug('Aborting scan in progress and starting new scan.')
+      await this.abort()
+    }
+
+    await this.state?.wait()
+    const unlock = await this.lock.lock()
+
+    try {
+      if (!start) {
+        start = await this.wallet.getEarliestHead()
+      }
+
+      if (!end) {
+        end = await this.wallet.getChainHead()
+      }
+
+      const chainProcessor = new RemoteChainProcessor({
+        logger: this.logger,
+        nodeClient: this.nodeClient,
+        maxQueueSize: this.maxQueueSize,
+        head: start?.hash ?? null,
+      })
+
+      chainProcessor.onAdd.on(async ({ header, transactions }) => {
+        await this.connectBlock(header, transactions, this.state?.abortController)
+        this.state?.signal(header)
+      })
+
+      chainProcessor.onRemove.on(async ({ header, transactions }) => {
+        await this.disconnectBlock(header, transactions, this.state?.abortController)
+        this.state?.signal(header)
+      })
+
+      if (start === null) {
+        start = await this.wallet.getChainGenesis()
+      }
+
+      this.logger.debug(`Scan starting from block ${start.sequence} to ${end.sequence}`)
+
+      const scan = new ScanState(start, end)
+      this.state = scan
+
+      void (async () => {
+        let hashChanged = true
+        while (hashChanged) {
+          const result = await chainProcessor.update({ signal: scan.abortController.signal })
+          hashChanged = result.hashChanged
```

Comment:
The old version of the scan updated `chainProcessor.hash` to `wallet.getEarliestHeadHash` on every loop. I think we may want to do this here or else accounts will be left behind when stopping, then starting scanning.

Reply:
Made this change